lms/djangoapps/gating/api.py

```diff
@@ -63,35 +61,60 @@ def _get_subsection_percentage(subsection_grade):
+
+
+def evaluate_entrance_exam(course_grade, user):
     """
     Evaluates any entrance exam milestone relationships attached
-    to the given subsection. If the subsection_grade meets the
-    minimum score required, the dependent milestone will be marked
+    to the given course. If the course_grade meets the
+    minimum score required, the dependent milestones will be marked
     fulfilled for the user.
     """
+    course = course_grade.course
     if milestones_helpers.is_entrance_exams_enabled() and getattr(course, 'entrance_exam_enabled', False):
-        subsection = modulestore().get_item(subsection_grade.location)
-        in_entrance_exam = getattr(subsection, 'in_entrance_exam', False)
-        if in_entrance_exam:
-            # We don't have access to the true request object in this context, but we can use a mock
-            request = RequestFactory().request()
-            request.user = user
-            exam_pct = get_entrance_exam_score(request, course)
-            if exam_pct >= course.entrance_exam_minimum_score_pct:
-                exam_key = UsageKey.from_string(course.entrance_exam_id)
+        if get_entrance_exam_content(user, course):
+            exam_chapter_key = get_entrance_exam_usage_key(course)
+            exam_percentage = get_entrance_exam_percentage(course_grade, exam_chapter_key)
+            if exam_percentage >= course.entrance_exam_minimum_score_pct:
                 relationship_types = milestones_helpers.get_milestone_relationship_types()
                 content_milestones = milestones_helpers.get_course_content_milestones(
                     course.id,
-                    exam_key,
+                    exam_chapter_key,
                     relationship=relationship_types['FULFILLS']
                 )
-                # Mark each milestone dependent on the entrance exam as fulfilled by the user.
+                # Mark each entrance exam dependent milestone as fulfilled by the user.
                 for milestone in content_milestones:
-                    milestones_helpers.add_user_milestone({'id': request.user.id}, milestone)
+                    milestones_helpers.add_user_milestone({'id': user.id}, milestone)
+
+
+def get_entrance_exam_usage_key(course):
+    """
+    Returns the UsageKey of the entrance exam for the course.
+    """
+    return UsageKey.from_string(course.entrance_exam_id).replace(course_key=course.id)
+
+
+def get_entrance_exam_percentage(course_grade, exam_chapter_key):
+    """
+    Returns the percentage value for the given chapter.
+    """
+    score = 0.0
+    chapter_grade = course_grade.chapter_grades.get(exam_chapter_key, None)
+    if chapter_grade:
+        earned = 0.0
+        possible = 0.0
+        for section in chapter_grade['sections']:
+            earned += section.graded_total.earned
+            possible += section.graded_total.possible
+        score = _calculate_percentage(earned, possible)
```

Comment:
As an example, this method is also claiming to return a percentage but is in fact returning a value in [0, 1]